internal/vdso.go

```diff
@@ -0,0 +1,133 @@
+package internal
+
+import (
+	"bytes"
+	"debug/elf"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strings"
+)
+
+const (
+	_AT_NULL         = 0
+	_AT_SYSINFO_EHDR = 33
+)
+
+// vdsoVersion returns the LINUX_VERSION_CODE embedded in the vDSO library linked to the current process.
+func vdsoVersion() (uint32, error) {
+	// read data from the auxiliary vector, which is normally passed directly to the process.
+	// Go does not expose that data, so we must read it from procfs.
+	// https://man7.org/linux/man-pages/man3/getauxval.3.html
+	av, err := os.Open("/proc/self/auxv")
+	if err != nil {
+		return 0, fmt.Errorf("open auxv: %w", err)
+	}
+	defer av.Close()
+
+	vdsoAddr, err := vdsoMemoryAddress(av)
+	if err != nil {
+		return 0, fmt.Errorf("vdso elf header: %w", err)
+	}
+
+	// use /proc/self/mem rather than unsafe.Pointer tricks
+	mem, err := os.Open("/proc/self/mem")
+	if err != nil {
+		return 0, fmt.Errorf("open mem: %w", err)
+	}
+	defer mem.Close()
+
+	// open ELF at provided memory address, as offset into /proc/self/mem
+	r := io.NewSectionReader(mem, int64(vdsoAddr), math.MaxInt64)
+	return linuxVersionCode(r)
+}
+
+// vdsoMemoryAddress returns the memory address of the vDSO library linked to the current process.
+func vdsoMemoryAddress(r io.Reader) (uint64, error) {
+	abuf, err := io.ReadAll(r)
+	if err != nil {
+		return 0, fmt.Errorf("readall auxv: %w", err)
+	}
+
+	br := bytes.NewReader(abuf)
+	// ensure we only read the exact number of uint64s
+	buf := make([]uint64, len(abuf)/8)
+	err = binary.Read(br, NativeEndian, &buf)
+	if err != nil {
+		return 0, fmt.Errorf("binary read auxv: %w", err)
+	}
+
+	// loop through all the type+value pairs until we find `AT_SYSINFO_EHDR`, described as:
+	// The address of a page containing the virtual Dynamic
+	// Shared Object (vDSO) that the kernel creates in order to
+	// provide fast implementations of certain system calls.
+	for i := 0; i < len(buf)-1 && buf[i] != _AT_NULL; i += 2 {
+		tag, val := buf[i], buf[i+1]
+		switch tag {
+		case _AT_SYSINFO_EHDR:
+			if val == 0 {
```

Comment:
So there may be multiple _AT_SYSINFO_EHDR? I would expect `0` to be an error or something.

Reply:
I don't think there will be multiple. As-is, it will result in the `not found` error, but after looping through the rest of the pairs.

Reply:
In that case I'd either skip the check and allow zero (it'll fail elsewhere i assume) or return a dedicated error here.

Reply:
Returning an error now.